Initial implementation of GPUComputePipeline

imiklos opened the pull request:
<!-- Please describe your changes on the following line: -->
Added WebIDL bindings for `GPUComputePipeline`.
Implemented the `createComputePipeline` function of `GPUDevice`.
This pull request is based on: #25677 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes addresses a part of #24706

cc @kvark @jdm @zakorgy
<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Commit:
Initial implementation of GPUShaderModule

Added WebIDL bindings for `GPUShaderModule`.
Implemented the `createShaderModule` function of `GPUDevice`.

Commit:
Initial implementation of GPUComputePipeline

Added WebIDL bindings for `GPUComputePipeline`.
Implemented the `createComputePipeline` function of `GPUDevice`.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/gpupipelinelayout.rs, components/script/dom/gpushadermodule.rs, components/script/dom/globalscope.rs, components/script/dom/identityhub.rs, components/script/dom/mod.rs and 6 more
 * @kichjang: components/script/dom/gpupipelinelayout.rs, components/script/dom/gpushadermodule.rs, components/script/dom/globalscope.rs, components/script/dom/identityhub.rs, components/script/dom/mod.rs and 6 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

Manishearth commented:
r? @kvark

kvark commented on components/script/dom/globalscope.rs:
```diff
@@ -2128,6 +2131,17 @@ impl GlobalScope {
+            .create_shader_module_id(backend)
+    }
+    pub fn wgpu_create_compute_pipeline_id(&self, backend: Backend) -> ComputePipelineId {
+        self.gpu_id_hub
```
A bit surprised you are able to use the `RefCell` here. In Gecko implementation, multiple threads in the same content process may want to create their own resource IDs.

jdm replied:
@imiklos Please file an issue about this - it sounds like when we create an ID, we could have multiple threads that end up creating the same ID independently. Possible we need to share a single gpu_id_hub between all threads using Arc & Mutex?

kvark commented on components/script/dom/globalscope.rs:
```diff
@@ -2128,6 +2131,17 @@ impl GlobalScope {
             .create_pipeline_layout_id(backend)
     }
+
+    pub fn wgpu_create_shader_module_id(&self, backend: Backend) -> ShaderModuleId {
```
nit: perhaps it would be simpler to just have a function returning `&mut GpuIdHub` and let the users call the methods appropriately?

kvark commented on components/script/dom/gpudevice.rs:
```diff
@@ -523,9 +529,74 @@ impl GPUDeviceMethods for GPUDevice {
+            String(program) => program
+                .as_bytes()
+                .iter()
+                .map(|c| *c as u32)
```
oh wow, how is this supposed to work? We are iterating characters of a string and casting them to `u32` words?

imiklos replied:
I wasn't sure about this one. I've updated the #25677 pull request addressing this in it.

kvark commented on components/script/dom/gpudevice.rs:
```diff
@@ -523,9 +529,74 @@ impl GPUDeviceMethods for GPUDevice {
+            ))
+            .expect("Failed to create WebGPU ShaderModule");
+
+        let shader_module = receiver.recv().unwrap();
```
In all of the creation methods, we aren't expected to block on anything. Here, therefore, we can't `recv()`. Instead, we'd be creating a new ID and asking the GPU server side to create something associated with this ID, but never wait for it to do so.

imiklos replied:
Yes, We have an issue ( #25472 ) for this, we would like to address this in a separate pull request for all the call which has to be non-blocking.

kvark commented on components/script/dom/gpudevice.rs:
```diff
@@ -523,9 +529,74 @@ impl GPUDeviceMethods for GPUDevice {
+            .as_bytes()
+            .iter()
+            .map(|c| *c as i8)
+            .collect::<Vec<i8>>();
```
why isn't this a `String`?

imiklos replied:
You are right! This should be a `String`.

kvark commented on components/webgpu/lib.rs:
```diff
@@ -286,6 +300,50 @@ impl WGPU {
+                        layout,
+                        compute_stage: wgpu_core::pipeline::ProgrammableStageDescriptor {
+                            module: program,
+                            entry_point: entry.as_ptr(),
```
what makes it null-terminated?

imiklos replied:
I've created a `CString` from the entry_point which will be null-terminated then passed it as pointer to the descriptor.

imiklos commented:
@kvark Thank you for the review! I've addressed what you mentioned, I'm still not sure about the `ShaderModule` creation in case when we are getting it as `String`. I've also rebased this pull request to the #25677.

kvark commented:
>I'm still not sure about the ShaderModule creation in case when we are getting it as String

We basically don't support this (yet). It depends on the shading language discussions...

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 9031369 has been approved by `jdm`

<!-- @bors-servo r=jdm 9031369c19f3274a1dee4e15e67c0c539d8aa976 -->

bors-servo commented:
:hourglass: Testing commit 9031369c19f3274a1dee4e15e67c0c539d8aa976 with merge e0883130d03296b3ef0d7c6d8189ae8e08e134e9...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/W0O_BbW7Qv2jdf-F05Awqg)

jdm commented:
@bors-servo retry
* #24726

bors-servo commented:
:hourglass: Testing commit 9031369c19f3274a1dee4e15e67c0c539d8aa976 with merge e89e78d401c785cd11189c5d448a6248165c5a83...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ObCck_A5TuWwNyUwReAbzQ)
Approved by: jdm
Pushing e89e78d401c785cd11189c5d448a6248165c5a83 to master...
